Add JOSS paper.

vnmabus opened the pull request:
## Describe the proposed changes
This adds a paper to the repo, so that we can submit it to JOSS.

@trossi I had to rewrite part of the text that I had, as it was for an older version (prior to pyOpenSci submission). It can probably (almost surely, I would say) be improved, in case you have suggestions to do so.
Other things to take into account:
- Please review your author info. I tried to add your ORCID and affiliation, but it may be incorrect.
- Please add your acknowledgements (funding or whatever you need).
- Please feel free to add the work in which you used the library at the end of the "Statement of need" section.
- I left the section "Ongoing work" (name can be changed) for you to explain the changes you did.
- You can compile locally using the Docker they provide (https://joss.readthedocs.io/en/latest/paper.html#docker). I also set up the Github action, to check the paper after you push (e.g. https://github.com/vnmabus/rdata/actions/runs/10648562090).

You can add your changes either as suggestions to this PR or as PRs against this branch.

vnmabus committed:
First version of the paper.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,130 @@
+---
+title: 'rdata: Read R datasets from Python'
```
```suggestion
title: 'rdata: A Python library for R datasets'
```
I'd suggest a broader title so that the title doesn't limit the scope to reading only. This would be useful in long term as the package develops, but the JOSS paper remains a fixed reference.

Possible alternative could be also: `rdata: A Python package for R datasets`.

Edit: added "A" to the titles.

vnmabus replied:
I agree with this, although I am not sure if the new title is maybe too broad. If I can think something better I will tell you.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,130 @@
+ - name: Universidad Autónoma de Madrid, Spain
+   index: 1
+ - name: CSC - IT Center for Science Ltd, Finland
+   index: 2
```
```suggestion
  - name: Universidad Autónoma de Madrid, Spain
    index: 1
  - name: CSC – IT Center for Science Ltd., Finland
    index: 2
```
Minor change in dash and aligning spacing with the rest of yaml.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,130 @@
+   index: 1
+ - name: CSC - IT Center for Science Ltd, Finland
+   index: 2
+date: 31 August 2024
```
```suggestion
date: 4 September 2024
```
As of today.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,130 @@
+# Summary
+
+Research work usually requires the analysis and processing of data from different sources.
+Traditionally statisticians and other research professionals have been using R for this task, and have compiled a huge amount of datasets in the Rda and Rds formats, native to this programming language.
```
```suggestion
Traditionally in statistical computing, R language has been widely used for this task, and a huge amount of datasets have been compiled in the Rda and Rds formats, native to this programming language.
```
Suggestion to rephrase as "and other research professionals" is covering all science although R is limited to statistical computing mainly (?).

vnmabus replied:
Yes, I agree. I didn't want to say just "statisticians", because there are users from other fields of science (but mostly in a statistical context). I think the way you phrased it is much better.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,130 @@
+Traditionally statisticians and other research professionals have been using R for this task, and have compiled a huge amount of datasets in the Rda and Rds formats, native to this programming language.
+As these formats contain internally the representation of R objects, they cannot be directly used from Python, another widely used language for data analysis and processing.
+The library `rdata` allows to load and convert these datasets to Python objects, without the need of exporting them to other intermediate formats which may not keep all the original information.
+This library has minimal dependencies, ensuring that it can be used in contexts where an R installation is not available.
```
```suggestion
This library has minimal dependencies, ensuring that it can be used in contexts where an R installation is not available.
The capability to write data in Rda and Rds formats is also under development.
```
Adding one sentence (the first sentence unchanged).

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,130 @@
+Secondly, launching R just to load data is inefficient, both in time and memory.
+Finally, this package inherits the GPL license from the R language, which is not compatible with most Python packages, typically released under more permissive licenses.
+
+The recent package `pyreadr` [@fajardo_2018_pyreadr] also provides functionality to read some R datasets.
```
```suggestion
The package `pyreadr` [@fajardo_2024_pyreadr] also provides functionality to read and write some R datasets.
```
I'd skip "recent" and mention writing here too. (Note also related suggestion in bib file.)

vnmabus replied:
Yeah, I agree. The "recent" adjective was just to point out that "rdata" predates "pyreadr" by several months, but I agree that it should not be of importance.

trossi commented on paper/paper.bib:
```diff
@@ -0,0 +1,53 @@
+  month = dec,
+  doi = {10.5281/zenodo.7110170},
+  url = {https://github.com/ofajardo/pyreadr}
+}
```
```suggestion
@misc{fajardo_2024_pyreadr,
  title = {Pyreadr},
  author = {Fajardo, Otto},
  year = {2024},
  month = jul,
  doi = {10.5281/zenodo.13132498},
  url = {https://github.com/ofajardo/pyreadr}
}
```
Updating the link to the latest release in zenodo.

By the way, other references seem to use the zenodo doi "to all versions". Maybe those could be updated to the fixed version variant or vice versa here?

Another question that applies to all references with zenodo doi: should all the years be updated to the latest release?

vnmabus replied:
I think I would rather use the Zenodo DOI "to all versions", unless there is a reason for referencing a particular version.

About the years of Zenodo DOIs, if we use the "to all versions" DOI I guess it is reasonable to update the year.

trossi replied:
>  I think I would rather use the Zenodo DOI "to all versions", unless there is a reason for referencing a particular version. 
>
> About the years of Zenodo DOIs, if we use the "to all versions" DOI I guess it is reasonable to update the year.

I'll suggest a set of changes accordingly.

trossi commented on paper/paper.bib:
```diff
@@ -0,0 +1,53 @@
+  abstract = {The library scikit-fda is a Python package for functional data analysis (FDA). It provides a comprehensive set of tools for representation, preprocessing, and exploratory analysis of functional data. The library is built upon and integrated in Python's scientific ecosystem. In particular, it conforms to the scikit-learn application programming interface so as to take advantage of th ...
+  copyright = {Copyright (c) 2024 Carlos Ramos-Carre{\~n}o, Jos{\'e} Luis Torrecilla, Miguel Carbajo-Berrocal, Pablo Marcos, Alberto Su{\'a}rez},
+  langid = {english}
+}
```
```suggestion
}

@article{rahman+_2024_hmschpc,
  title = {Accelerating joint species distribution modelling with {Hmsc-HPC} by {GPU} porting},
  author = {Rahman, Anis Ur and Tikhonov, Gleb and Oksanen, Jari and Rossi, Tuomas and Ovaskainen, Otso},
  year = {2024},
  month = sep,
  journal = {PLOS Computational Biology},
  volume = {20},
  number = {9},
  pages = {e1011914},
  doi = {10.1371/journal.pcbi.1011914},
  abstract = {Joint species distribution modelling (JSDM) is a widely used statistical method that analyzes combined patterns of all species in a community, linking empirical data to ecological theory and enhancing community-wide prediction tasks. However, fitting JSDMs to large datasets is often computationally demanding and time-consuming. Recent studies have introduced new statistical and machine learning techniques to provide more scalable fitting algorithms, but extending these to complex JSDM structures that account for spatial dependencies or multi-level sampling designs remains challenging. In this study, we aim to enhance JSDM scalability by leveraging high-performance computing (HPC) resources for an existing fitting method. Our work focuses on the Hmsc R-package, a widely used JSDM framework that supports the integration of various dataset types into a single comprehensive model. We developed a GPU-compatible implementation of its model-fitting algorithm using Python and the TensorFlow library. Despite these changes, our enhanced framework retains the original user interface of the Hmsc R-package. We evaluated the performance of the proposed implementation across various model configurations and dataset sizes. Our results show a significant increase in model fitting speed for most models compared to the baseline Hmsc R-package. For the largest datasets, we achieved speed-ups of over 1000 times, demonstrating the substantial potential of GPU porting for previously CPU-bound JSDM software. This advancement opens promising opportunities for better utilizing the rapidly accumulating new biodiversity data resources for inference and prediction.},
}
```

Reference to our use case.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,130 @@
+The license can also be a problem, as it is part of the GPL family and does not allow commercial use.
+
+As existing solutions were unsuitable for our needs, the package `rdata` was developed to parse data in the RData format.
+This is a small, extensible and very complete implementation in pure Python of a RData parser, that is able to read and convert most datasets in the CRAN repository to equivalent Python objects.
```
```suggestion
This is a small, extensible, efficient, and very complete implementation in pure Python of a RData parser, that is able to read and convert most datasets in the CRAN repository to equivalent Python objects.
```
Suggestion to add "efficient" for #32 .

Out of curiosity, I run a quick test comparing reading times to `pyreadr`:

```bash
# Generate test files with and without compression
for i in {1..7}; do n=$(( 2 ** $i )); Rscript -e "saveRDS(runif(n=$n*1024**2), file='array_$i.rds', compress=FALSE)"; done
for i in {1..7}; do n=$(( 2 ** $i )); Rscript -e "saveRDS(runif(n=$n*1024**2), file='array_${i}_comp.rds')"; done

# Read with rdata and pyreadr
for i in {7..1}; do echo $i; time -p python -c "from rdata import read_rds; read_rds('array_$i.rds')"; done
for i in {7..1}; do echo $i; time -p python -c "from pyreadr import read_r; read_r('array_$i.rds')"; done
for i in {7..1}; do echo $i; time -p python -c "from rdata import read_rds; read_rds('array_${i}_comp.rds')"; done
for i in {7..1}; do echo $i; time -p python -c "from pyreadr import read_r; read_r('array_${i}_comp.rds')"; done
```

Data on a laptop (more sampling and fixed environment would be needed for more reliable numbers):
```raw
# 1. column: i
# 2. column: for i in {7..1}; do echo $i; time -p python -c "from rdata import read_rds; read_rds('array_$i.rds')"; done
# 3. column: for i in {7..1}; do echo $i; time -p python -c "from pyreadr import read_r; read_r('array_$i.rds')"; done
# 4. column: for i in {7..1}; do echo $i; time -p python -c "from rdata import read_rds; read_rds('array_${i}_comp.rds')"; done
# 5. column: for i in {7..1}; do echo $i; time -p python -c "from pyreadr import read_r; read_r('array_${i}_comp.rds')"; done
7  2.69   8.25   7.79   11.02
6  0.96   4.28   3.14   5.63
5  0.65   2.27   1.74   3.00
4  0.59   1.31   1.02   1.67
3  0.46   0.84   0.69   1.01
2  0.40   0.59   0.52   0.68
1  0.37   0.47   0.44   0.52
```

Plot:
![image](https://github.com/user-attachments/assets/4adb24ed-7d2f-4c9b-ac98-ea8b6828316f)

vnmabus replied:
Indeed, it seems that efficient is appropriate! I will add the suggestion.  However, note that these examples compare only the performance parsing one homogeneous array. It could be the case that for more complex datasets containing large hierarchies of R objects the comparison has different results. As analyzing that is a complex problem, I would rather not add the comparisons to the paper (but thanks for carrying them out, I was curious too).

trossi replied:
I agree. I focused on such large arrays as they are most important for our use case, but general case is different more complicated. I wouldn't add any of such comparisons to the paper either. I feel that they would be out of scope too as the main advantage of rdata is extensibility and completeness of the type support (in my opinion).

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,130 @@
+
+# Ongoing work
+
+
```
```suggestion
To broaden the utility of the `rdata` library to data processing pipelines with steps in both R and Python, we are currently extending the library with the capability to write compatible Python objects to RData files.
As an example, such a pipeline is present in the Hmsc-HPC code [@rahman+_2024_hmschpc], the continuous development of which has been driving the ongoing work on the writing functionality in the `rdata` library.
The writing of RData files is implemented as a two-step process similar to reading: first, the Python object is converted to the tree-like intermediate representation used in parsing, and then this intermediate representation is written to a RData file.
Currently, the writing functionality supporting common types is available at the development branch of the `rdata` library.
```

Here is a short description and context (Hmsc-HPC) for my contributions. Let me know if you think that the context description would fit better under 'Statement of need'.

vnmabus replied:
I think here fits too.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,130 @@
+
+# Acknowledgements
+
+The authors acknowledge financial support from the Spanish Ministry of Education and Innovation, projects PID2019-106827GB-I00 / AEI / 10.13039/501100011033 and PID2019-109387GB-I00.
+This work was also supported by an FPU grant (Formación de Profesorado Universitario) from the Spanish Ministry of Science, Innovation and Universities(MICIU) with reference FPU18/00047.
```
```suggestion
This work has received funding
from the Spanish Ministry of Education and Innovation, projects PID2019-106827GB-I00 / AEI / 10.13039/501100011033 and PID2019-109387GB-I00,
from an FPU grant (Formación de Profesorado Universitario) from the Spanish Ministry of Science, Innovation and Universities(MICIU) with reference FPU18/00047,
and from the European Union's Horizon Europe research and innovation programme under grant agreement No 101057437 (BioDT project, [https://doi.org/10.3030/101057437](https://doi.org/10.3030/101057437)).
Views and opinions expressed are those of the author(s) only and do not necessarily reflect those of the European Union or the European Commission. Neither the European Union nor the European Commission can be held responsible for them.
```

I added my acknowledgement and a suggestion to combine the statements to condense a bit. If you prefer, my acknowledgement could be a separate sentence too ("This work has also received funding from the European ...").

(And sorry for the lengthy disclaimer that comes with Horizon Europe projects.)

vnmabus replied:
Thanks, I think having them all in one sentence as you suggest is probably better.

trossi reviewed (commented):
@vnmabus Thank you! I have added a description of the writing functionality and some general suggestions.

vnmabus reviewed (commented):
I agree with all your suggestions. I commit them and will re-read the article to see if more improvements need to be made. I suggest you to do the same if/when you have time.

vnmabus committed:
Apply suggestions from code review

Co-authored-by: Tuomas Rossi <34502776+trossi@users.noreply.github.com>

trossi commented on paper/paper.bib:
```diff
@@ -0,0 +1,66 @@
+  author = {Fajardo, Otto},
+  year = {2024},
+  month = jul,
+  doi = {10.5281/zenodo.13132498},
```
```suggestion
  publisher = {Zenodo},
  doi = {10.5281/zenodo.7110169},
```

Here is the doi to "all versions" and adding Zenodo as publisher as for pandas.

trossi commented on paper/paper.bib:
```diff
@@ -0,0 +1,66 @@
+@misc{diaz-vico+ramos-carreno_2022_scikitdatasets,
+  title = {{{scikit-datasets}}: {{Scikit-learn-compatible}} Datasets},
+  author = {{D{\'i}az-Vico}, David and {Ramos-Carre{\~n}o}, Carlos},
+  year = {2022},
+  month = mar,
```
```suggestion
@misc{diaz-vico+ramos-carreno_2023_scikitdatasets,
  title = {{{scikit-datasets}}: {{Scikit-learn-compatible}} Datasets},
  author = {{D{\'i}az-Vico}, David and {Ramos-Carre{\~n}o}, Carlos},
  year = {2023},
  month = aug,
```

Updating the time stamp. (DOI is already to "all versions".)

trossi commented on paper/paper.bib:
```diff
@@ -0,0 +1,66 @@
+  url = {https://github.com/rpy2/rpy2}
+}
+
+@software{pandasdevelopmentteam_2020_pandasdev,
+  title = {{{pandas-dev/pandas}}: {{pandas}}},
+  author = {Pandas Development Team},
+  year = {2020},
+  month = feb,
```
```suggestion
@software{pandasdevelopmentteam_2024_pandasdev,
  title = {{{pandas-dev/pandas}}: {{pandas}}},
  author = {{The Pandas Development Team}},
  year = {2024},
  month = apr,
```

Updating date and fixing the rendering of author. DOI is already to "all versions".

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+It has a permissive license and can be extended to support additional conversions from custom R classes.
+
+The package `rdata` has been designed as a pure Python package with minimal dependencies, so that it can be easily integrated inside other libraries and applications.
+It currently powers the functionality offered in the `scikit-datasets` package [@diaz-vico+ramos-carreno_2022_scikitdatasets] for loading datasets from the CRAN repository of R packages.
```
```suggestion
It currently powers the functionality offered in the `scikit-datasets` package [@diaz-vico+ramos-carreno_2023_scikitdatasets] for loading datasets from the CRAN repository of R packages.
```

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+
+Advanced users will probably require loading datasets which contain non standard S3 or S4 classes, translating each of them to a custom Python class.
+This is easy to achieve using `rdata` by simply creating a constructor function that receives the converted object representation and its attributes, and returns a Python object of the desired type.
+As an example, consider the following simple code that constructs a `Pandas` [@pandasdevelopmentteam_2020_pandasdev] `Categorical` object from the internal representation of an R `factor`.
```
```suggestion
As an example, consider the following simple code that constructs a `Pandas` [@pandasdevelopmentteam_2024_pandasdev] `Categorical` object from the internal representation of an R `factor`.
```

trossi commented:
> I agree with all your suggestions. I commit them and will re-read the article to see if more improvements need to be made. I suggest you to do the same if/when you have time.

Thank you! I added the changes related to discussion on `paper.bib`. I re-read the article and I think it's in good state. I can have another look later this week.

trossi commented on paper/paper.bib:
```diff
@@ -0,0 +1,66 @@
+  publisher = {GitHub},
+  url = {https://github.com/rpy2/rpy2}
+}
+
```
```suggestion

@article{harris+_2020_numpy,
  title = {Array programming with {NumPy}},
  author = {Charles R. Harris and K. Jarrod Millman and St{\'{e}}fan J.
            van der Walt and Ralf Gommers and Pauli Virtanen and David
            Cournapeau and Eric Wieser and Julian Taylor and Sebastian
            Berg and Nathaniel J. Smith and Robert Kern and Matti Picus
            and Stephan Hoyer and Marten H. van Kerkwijk and Matthew
            Brett and Allan Haldane and Jaime Fern{\'{a}}ndez del
            R{\'{i}}o and Mark Wiebe and Pearu Peterson and Pierre
            G{\'{e}}rard-Marchant and Kevin Sheppard and Tyler Reddy and
            Warren Weckesser and Hameer Abbasi and Christoph Gohlke and
            Travis E. Oliphant},
  year = {2020},
  month = sep,
  journal = {Nature},
  volume = {585},
  number = {7825},
  pages = {357--362},
  doi = {10.1038/s41586-020-2649-2},
}

@inproceedings{mckinney_2010_pandas,
  author = {{W}es {M}c{K}inney},
  title = {{D}ata {S}tructures for {S}tatistical {C}omputing in {P}ython},
  booktitle = {{P}roceedings of the 9th {P}ython in {S}cience {C}onference},
  pages = {56 - 61},
  year = {2010},
  editor = {{S}t\'efan van der {W}alt and {J}arrod {M}illman},
  doi = {10.25080/Majora-92bf1922-00a},
}

```

I'd suggest to cite also NumPy and Pandas papers as suggested in https://numpy.org/citing-numpy/ and https://pandas.pydata.org/about/citing.html .

vnmabus replied:
I agree.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+The license can also be a problem, as it is part of the GPL family and does not allow commercial use.
+
+As existing solutions were unsuitable for our needs, the package `rdata` was developed to parse data in the RData format.
+This is a small, extensible, efficient, and very complete implementation in pure Python of a RData parser, that is able to read and convert most datasets in the CRAN repository to equivalent Python objects.
```
```suggestion
This is a small, extensible, efficient, and very complete implementation in pure Python of a RData parser, that is able to read and convert most datasets in the CRAN repository to equivalent Python objects, such as the built-in types of The Python Standard Library, NumPy arrays [@harris+_2020_numpy], or Pandas dataframes [@mckinney_2010_pandas; @pandasdevelopmentteam_2024_pandasdev].
```

Referring to the NumPy and Pandas papers.

trossi reviewed (commented):
I added a small suggestion including references to NumPy and Pandas publications.

vnmabus committed:
Apply suggestions from code review

Accepted changes from code review:
- Updated references
- Set Zenodo DOIs to "all versions" DOI
- Add references to NumPy and Pandas

Co-authored-by: Tuomas Rossi <34502776+trossi@users.noreply.github.com>

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+# Summary
+
+Research work usually requires the analysis and processing of data from different sources.
+Traditionally in statistical computing, R language has been widely used for this task, and a huge amount of datasets have been compiled in the Rda and Rds formats, native to this programming language.
```
```suggestion
Traditionally in statistical computing, the R language has been widely used for this task, and a huge amount of datasets have been compiled in the Rda and Rds formats, native to this programming language.
```

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+The library `rdata` allows to load and convert these datasets to Python objects, without the need of exporting them to other intermediate formats which may not keep all the original information.
+This library has minimal dependencies, ensuring that it can be used in contexts where an R installation is not available.
+The capability to write data in Rda and Rds formats is also under development.
+Thus, the library `rdata` facilitates data interchange, enabling the usage of the same datasets in both languages (e.g. for reproducibility, comparisons of results against methods in both languages, or migration of processing pipelines to Python).
```
I would say that with the new writing functionality the following is possible, and it sounds better for R programmers (the way it was written before looked like R is no longer desirable, which may be true in some contexts, but not all of them):

```suggestion
Thus, the library `rdata` facilitates data interchange, enabling the usage of the same datasets in both languages (e.g. for reproducibility, comparisons of results against methods in both languages, or the creation of complex processing pipelines that involve steps in both R and Python).
```

What do you think, @trossi ?

trossi replied:
This is much better indeed. Thanks!

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+
+# Statement of need
+
+The datasets from the CRAN repository are stored in the R specific format RData.
+In Python, there were a few packages that could parse this file format, albeit all of them presented some limitations.
```
In the summary we talked about both Rda and Rds, so maybe we should do the same here. I think I wrote this before I had the ability to read Rds.
```suggestion
The datasets from the CRAN repository are stored in the R specific formats Rda and Rds.
In Python, there were a few packages that could parse these file formats, albeit all of them presented some limitations.
```

trossi replied:
Sound good! I had interpreted that RData is an overall name covering both rda and rds files as they are almost the same, but I agree it's clearer to be explicit and mention both. I added suggestions for the remaining mentions of "RData".

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+In Python, there were a few packages that could parse this file format, albeit all of them presented some limitations.
+
+The package `rpy2` [@gautier_2024_rpy2] can be used to interact with R from Python.
+This includes the ability to load data in the RData format, and to convert these data to equivalent Python objects.
```
```suggestion
This includes the ability to load data in the Rda and Rds formats, and to convert these data to equivalent Python objects.
```

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+
+The package `rpy2` [@gautier_2024_rpy2] can be used to interact with R from Python.
+This includes the ability to load data in the RData format, and to convert these data to equivalent Python objects.
+Although this is arguably the best package to achieve interaction between both languages, it has many disadvantages if one wants to use it just to load RData datasets.
```
```suggestion
Although this is arguably the best package to achieve interaction between both languages, it has many disadvantages if one wants to use it just to load R datasets.
```

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+does not include the parsing of common objects such as R lists and S4 objects.
+The license can also be a problem, as it is part of the GPL family and does not allow commercial use.
+
+As existing solutions were unsuitable for our needs, the package `rdata` was developed to parse data in the RData format.
```
```suggestion
As existing solutions were unsuitable for our needs, the package `rdata` was developed to parse data in the Rda and Rds formats.
```

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+The license can also be a problem, as it is part of the GPL family and does not allow commercial use.
+
+As existing solutions were unsuitable for our needs, the package `rdata` was developed to parse data in the RData format.
+This is a small, extensible, efficient, and very complete implementation in pure Python of a RData parser, that is able to read and convert most datasets in the CRAN repository to equivalent Python objects, such as the built-in types of The Python Standard Library, NumPy arrays [@harris+_2020_numpy], or Pandas dataframes [@mckinney_2010_pandas; @pandasdevelopmentteam_2024_pandasdev].
```
```suggestion
This is a small, extensible, efficient, and very complete implementation in pure Python of an R data parser, that is able to read and convert most datasets in the CRAN repository to equivalent Python objects, such as the built-in types of the Python standard library, NumPy arrays [@harris+_2020_numpy], or Pandas dataframes [@mckinney_2010_pandas; @pandasdevelopmentteam_2024_pandasdev].
```

trossi replied:
Looks good, but suggesting to change to "an R data parser" because "R" has vowel sound.

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+The package `rdata` is intended to be both flexible and easy to use.
+In order to be flexible, the parsing of the RData format and the conversion of the parsed structures to appropriate Python objects have been splitted in two steps.
+This allows advanced users to perform custom conversions without losing information.
+Most users, however, will want to use the default conversion routine, which attempts to convert data
+to a standard Python representation which preserves most part of the information.
```
```suggestion
Most users, however, will want to use the default conversion routine, which attempts to convert data to a standard Python representation which preserves most part of the information.
Converting an Rda dataset to Python objects using the package `rdata` can be easily done as follows:
```

trossi replied:
Suggesting to change the article to "an Rda dataset" here too.

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+To broaden the utility of the `rdata` library to data processing pipelines with steps in both R and Python, we are currently extending the library with the capability to write compatible Python objects to RData files.
+As an example, such a pipeline is present in the Hmsc-HPC code [@rahman+_2024_hmschpc], the continuous development of which has been driving the ongoing work on the writing functionality in the `rdata` library.
+The writing of RData files is implemented as a two-step process similar to reading: first, the Python object is converted to the tree-like intermediate representation used in parsing, and then this intermediate representation is written to a RData file.
+Currently, the writing functionality supporting common types is available at the development branch of the `rdata` library.
```
Shouldn't it be "in"?

```suggestion
Currently, the writing functionality supporting common types is available in the development branch of the `rdata` library.
```

vnmabus reviewed (commented):
I just re-read the article again and added some suggestions. Please @trossi give me your opinion on these when you have time.

trossi commented on paper/paper.md:
````diff
@@ -0,0 +1,137 @@
+converted
+```
+
+This is equivalent to the following code, in which the two steps have been performed separatedly.
````
```suggestion
This is equivalent to the following code, in which the two steps are performed separately:
```
Adding colon similarly to previous sentence.

trossi commented on paper/paper.md:
````diff
@@ -0,0 +1,137 @@
+converted = rdata.conversion.convert(parsed)
+```
+
+The function `parse_file()` of the parser module is used to parse the RData file, returning a tree-like structure of Python objects that contains a representation of the basic R objects conforming the dataset.
````
```suggestion
The function `parse_file()` of the parser module is used to parse Rda and Rds files, returning a tree-like structure of Python objects that contains a representation of the basic R objects conforming the dataset.
```

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+The function `convert()` of the conversion module transforms that representation to the final Python objects, such as lists, dictionaries or dataframes, that users can manipulate.
+
+Advanced users will probably require loading datasets which contain non standard S3 or S4 classes, translating each of them to a custom Python class.
+This is easy to achieve using `rdata` by simply creating a constructor function that receives the converted object representation and its attributes, and returns a Python object of the desired type.
+As an example, consider the following simple code that constructs a `Pandas` [@pandasdevelopmentteam_2024_pandasdev] `Categorical` object from the internal representation of an R `factor`.
```
```suggestion
This can be achieved using `rdata` by creating a constructor function that receives the converted object representation and its attributes, and returns a Python object of the desired type.
As an example, consider the following short code that constructs a `Pandas` [@pandasdevelopmentteam_2024_pandasdev] `Categorical` object from the internal representation of an R `factor`.
```
I would avoid word "simple" too much as it can be quite time-consuming to get things right (a short piece of code might not be that simple).

vnmabus replied:
I agree. Also in general I think it is better to not make such qualitative assessments in a paper, as difficulty is different for each person.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+If there are no constructors for any of the parent classes, the basic underlying Python object will be left without transformation.
+
+By default, a dictionary named `DEFAULT_CLASS_MAP` is passed to `convert()` including constructors for commonly used classes, such as `data.frame`, `ordered` or the aforementioned `factor`.
+In case anyone wants different conversions for basic R objects, it would be enough to create a subclass of the `Converter` class.
```
```suggestion
In case the user desires different conversions for basic R objects, it would be enough to create a subclass of the `Converter` class.
```

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+
+# Ongoing work
+
+To broaden the utility of the `rdata` library to data processing pipelines with steps in both R and Python, we are currently extending the library with the capability to write compatible Python objects to RData files.
```
```suggestion
To broaden the utility of the `rdata` library for data processing pipelines with steps in both R and Python, we are currently extending the library with the capability to write compatible Python objects to Rda and Rds files.
```

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+
+To broaden the utility of the `rdata` library to data processing pipelines with steps in both R and Python, we are currently extending the library with the capability to write compatible Python objects to RData files.
+As an example, such a pipeline is present in the Hmsc-HPC code [@rahman+_2024_hmschpc], the continuous development of which has been driving the ongoing work on the writing functionality in the `rdata` library.
+The writing of RData files is implemented as a two-step process similar to reading: first, the Python object is converted to the tree-like intermediate representation used in parsing, and then this intermediate representation is written to a RData file.
```
```suggestion
The writing of Rda and Rds files is implemented as a two-step process similar to reading: first, the Python object is converted to the tree-like intermediate representation used in parsing, and then this intermediate representation is written to a file of the chosen format.
```

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+# Ongoing work
+
+To broaden the utility of the `rdata` library to data processing pipelines with steps in both R and Python, we are currently extending the library with the capability to write compatible Python objects to RData files.
+As an example, such a pipeline is present in the Hmsc-HPC code [@rahman+_2024_hmschpc], the continuous development of which has been driving the ongoing work on the writing functionality in the `rdata` library.
```
```suggestion
As an example, such a pipeline is present in the Hmsc-HPC code [@rahman+_2024_hmschpc], and the continuous development of this code has been driving the ongoing work on the writing functionality of the `rdata` library.
```

vnmabus replied:
I feel this is better written as two sentences.

vnmabus replied:
```suggestion
As an example, such a pipeline is present in the Hmsc-HPC code [@rahman+_2024_hmschpc].
The continuous development of this code has been driving the ongoing work on the writing functionality of the `rdata` library.
```

trossi replied:
```suggestion
As an example, such a pipeline is present in the Hmsc-HPC code [@rahman+_2024_hmschpc].
The continuous development of this code has also been driving the ongoing work on the writing functionality of the `rdata` library.
```

I agree it's more readable. I'd further suggest to add "also" like here for bonding the sentences.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+# Features
+
+The package `rdata` is intended to be both flexible and easy to use.
+In order to be flexible, the parsing of the RData format and the conversion of the parsed structures to appropriate Python objects have been splitted in two steps.
```
```suggestion
In order to be flexible, the parsing of the R data file formats and the conversion of the parsed structures to appropriate Python objects have been splitted in two steps.
```

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+Finally, this package inherits the GPL license from the R language, which is not compatible with most Python packages, typically released under more permissive licenses.
+
+The package `pyreadr` [@fajardo_2024_pyreadr] also provides functionality to read and write some R datasets.
+It relies in the C library `librdata` in order to perform the parsing of the RData format.
```
```suggestion
It relies in the C library `librdata` in order to perform the parsing of the R data files.
```

trossi commented on paper/paper.md:
````diff
@@ -0,0 +1,137 @@
+As an example, consider the following simple code that constructs a `Pandas` [@pandasdevelopmentteam_2024_pandasdev] `Categorical` object from the internal representation of an R `factor`.
+
+```python
+import pandas
+
+
+def factor_constructor(obj, attrs):
+    values = [attrs['levels'][i - 1] if i >= 0 else None for i in obj]
+
+    return pandas.Categorical(values, attrs['levels'], ordered=False)
````
```suggestion
import pandas

def factor_constructor(obj, attrs):
    values = [attrs['levels'][i - 1] if i >= 0 else None for i in obj]
    return pandas.Categorical(values, attrs['levels'], ordered=False)
```

I'd suggest to remove these two empty lines for better readability in the rendered pdf.

trossi commented on paper/paper.md:
```diff
@@ -0,0 +1,137 @@
+import rdata
+
+converted = rdata.read_rda("dataset.rda")
+converted
```
```suggestion
```
I'd suggest to remove this line (or use `print(converted)` instead if printing is intended.

trossi reviewed (commented):
@vnmabus I agree with all your suggestions. I added a few comments and small suggestions.

vnmabus committed:
Apply suggestions from code review

Co-authored-by: Tuomas Rossi <34502776+trossi@users.noreply.github.com>

vnmabus committed:
Update paper/paper.md

Co-authored-by: Tuomas Rossi <34502776+trossi@users.noreply.github.com>

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,135 @@
+from an FPU grant (Formación de Profesorado Universitario) from the Spanish Ministry of Science, Innovation and Universities(MICIU) with reference FPU18/00047,
+and from the European Union's Horizon Europe research and innovation programme under grant agreement No 101057437 (BioDT project, [https://doi.org/10.3030/101057437](https://doi.org/10.3030/101057437)).
+Views and opinions expressed are those of the author(s) only and do not necessarily reflect those of the European Union or the European Commission. Neither the European Union nor the European Commission can be held responsible for them.
+
```
Adding missing acknowledgement.

```suggestion
The authors gratefully acknowledge the use of the computational facilities of Centro de Computación Científica (CCC) at Universidad Autónoma de Madrid.

```

trossi replied:
```suggestion
The authors gratefully acknowledge the use of the computational facilities provided by Centro de Computación Científica (CCC) at Universidad Autónoma de Madrid and by CSC – IT Center for Science, Finland.

```

I'd also like to add similar acknowledgement to [our facilities](https://docs.csc.fi/support/faq/how-to-cite-csc/). I also changed "of" to "provided by" to make the sentence structure work with both acknowledgements.

vnmabus committed:
Update paper/paper.md

Co-authored-by: Tuomas Rossi <34502776+trossi@users.noreply.github.com>

vnmabus commented on paper/paper.md:
```diff
@@ -0,0 +1,136 @@
+
+# Statement of need
+
+The datasets from the CRAN repository are stored in the R specific formats Rda and Rds.
```
I re-read the document. I think this sentence should be changed to place less emphasis on CRAN (which should be just an example). Apart from that I do not see anything else that I would like to change, so for my part it would be ready to send to JOSS after this change. What do you think?

```suggestion
The datasets of the R programming language, such as those from the CRAN repository, are stored in the R specific formats Rda and Rds.
```

trossi replied:
```suggestion
The datasets of the R programming language, such as those from the CRAN repository, are often stored in the R specific formats Rda and Rds.
```

The sentence sounds like only these format could be used in R, so I'd suggest to add "often".

I agree that the paper is ready to be submitted to JOSS.

vnmabus committed:
Update paper/paper.md

Co-authored-by: Tuomas Rossi <34502776+trossi@users.noreply.github.com>

vnmabus commented:
@trossi , I just sent the submission to JOSS.

trossi commented:
@vnmabus Great! Thank you!

vnmabus commented:
@trossi In case you need it for your CV: it seems that the article was fast-tracked... well, really fast 😅️: https://joss.theoj.org/papers/10.21105/joss.07540

I will merge this and make the necessary edits in the repo in the following days.

trossi commented:
@vnmabus Great, thank you!
